src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/DoubleBufferedStreamReader.cs

```diff
@@ -0,0 +1,169 @@
+            {
+                return this.ReadByteSlow();
+            }
+            else
+            {
+                this.position++;
+                return this.chunk[this.bytesRead++];
+            }
+        }
+
+        /// <summary>
+        /// Skips the number of bytes in the stream
+        /// </summary>
+        /// <param name="count">The number of bytes to skip</param>
+        public void Skip(int count)
+        {
+            this.Position += count;
+        }
+
+        /// <summary>
+        /// Reads a sequence of bytes from the current stream and advances the position within the stream
+        /// by the number of bytes read.
+        /// </summary>
+        /// <param name="buffer">
+        /// An array of bytes. When this method returns, the buffer contains the specified
+        /// byte array with the values between offset and (offset + count - 1) replaced by
+        /// the bytes read from the current source.
+        /// </param>
+        /// <param name="offset">
+        /// The zero-based byte offset in buffer at which to begin storing the data read
+        /// from the current stream.
+        /// </param>
+        /// <param name="count">The maximum number of bytes to be read from the current stream.</param>
+        /// <returns>
+        /// The total number of bytes read into the buffer. This can be less than the number
+        /// of bytes requested if that many bytes are not currently available, or zero (0)
+        /// if the end of the stream has been reached.
+        /// </returns>
+        public int Read(byte[] buffer, int offset, int count)
+        {
+            int n = 0;
+            if (buffer.Length <= ChunkLength)
+            {
+                if (this.position == 0 || count + this.bytesRead > ChunkLength)
+                {
+                    // Refill our buffer then copy.
+                    this.stream.Seek(this.position, SeekOrigin.Begin);
+                    this.stream.Read(this.chunk, 0, ChunkLength);
+                    this.bytesRead = 0;
+                }
+
+                Buffer.BlockCopy(this.chunk, this.bytesRead, buffer, offset, count);
+                this.position += count;
+                this.bytesRead += count;
+
+                n = Math.Min(count, (int)(this.Length - this.position));
+            }
+            else
+            {
+                // Read to target but don't copy to our chunk.
+                this.stream.Seek(this.position, SeekOrigin.Begin);
+                n = this.stream.Read(buffer, offset, count);
+
+                this.Position += count;
+            }
+
+            return Math.Max(n, 0);
+        }
+
+        /// <inheritdoc/>
+        public void Dispose()
+        {
+            this.buffer?.Dispose();
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private int ReadByteSlow()
+        {
+            this.stream.Seek(this.position, SeekOrigin.Begin);
```

Comment:
Wonder why do we need to seek the stream on every read?

Reply:
It was easier than tracking both positions. We could compare I guess then only seek if required.